keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md

```diff
@@ -645,7 +705,7 @@ a different pod the next time (or none). This would leave a pod with a
 condition that it was preempted, when it actually wasn't. This in turn
 could lead to inproper handling of the pod by the job controller.
```

Comment:
What if job controller looks at it before "disruption controller" cleans the condition?

Reply:
Pods remain transparent to the Job controller until they are in a terminal state (Succeeded or Failed) or are scheduled for deletion (deletionTimestamp != nil). So a pod, with a stale disruption condition will either:
1. continue running, being transparent to the job controller, until the disruption condition is cleaned
2. continue running till completion (potentially fail) before the disruption controller removes the stale condition. Then, the stale condition may remain and be checked by the job controller.

Reply:
Ok. But if the pod remains transparent for the Job controller, what's the point of cleaning the condition? Do we consider hypothetical users of the conditions other than the Job controller? I guess it could be considered a little cleaner, but it would still be a race condition trap for those other users to rely only on the condition without deletionTimestamp, right? Couldn't we say that the condition is only relevant/reliable/usable if deletionTimestamp is set, and therefore we don't need to clean it up?

Reply:
What's the concern with cleaning it up? This is already implementing it btw.

Also, the Pod might eventually finish by it's own means (and get a deletionTimestamp). At this point, the PodCondition might not be accurate.

Reply:
Just the concern around unnecessary code (not a huge concern, but trying to understand). It's always something to maintain, so I'd prefer to not have code without purpose. :)

If the purpose is for pods that finish on its own - fair enough. But additional concern that it's masking the race conditions that are still there, and creates false impression that it's okay to rely on the condition. E.g. for the case you mentioned, I guess it's still possible that the pod finishes on its own before the disruption controller cleans the stale condition?

(Btw. is the disruption controller needed for other things, or just this?)

Reply:
That is correct, in the mentioned scenario, the condition is not accurate. We can add this to the "risks and mitigations" section.

On the bright side, it's unlikely to happen. The sequence of events is:
1. a controller decides that a pod should be disrupted
2. the controller adds the condition successfully
3. the Pod deletion call fails
4. the controller restarts and doesn't think that this pod should be disrupted again

Having the removal is a single place (the disruption controller) makes it so that we only have to deal with the logic once. Otherwise we have to go into each controller, watch pods and delete the condition if the pod was not marked as to-be-disrupted somewhere in an in-memory cache. This sounds like overkill.

Reply:
I have added the risks and mitigation section: "Stale DisruptionTarget condition which is not clean up"